Blueshift: send id and/or anonymous_id, add name & category to page event

ruemic opened the pull request.

ruemic committed:
send customer_id & anonymous_id with identify,track,page,alias. send category/name with page

sperand-io commented on lib/blueshift/index.js:
```diff
@@ -52,32 +52,37 @@ Blueshift.prototype.page = function(page){
-  var traits = group.traits({ created: 'created_at' });
+Blueshift.prototype.identify = function(identify){
+  if (!identify.userId() || !identify.anonymousId()) {
+    this.debug('user id required');
```
missing early `return`

ruemic replied:
Ah, thnx @sperand-io 
pushed a fix

ruemic replied:
Hey sorry, I think I jumped the gun here.
We do actually want to fire an event even if there's no userId, so I've removed the early return again. 

Should I remove the `!identify.anonymousId()` part of the predicate?
It seems like it would always be present.

sperand-io commented on lib/blueshift/test.js:
```diff
@@ -94,57 +94,77 @@ describe('Blueshift', function(){
+          id: 'id',
+          customer_id: 'id',
+          _bsft_source: 'segment.com',
+          anonymousId: analytics._user.anonymousId()
```
`analytics.user().anonymousId()` :)

sperand-io commented:
@ruemic that was fast — thanks! i think switching all the refs to `analytics._user` to `analytics.user()` in the tests should do the trick.

after you get them passing, if you could please rebase and squash to a single commit that would be lovely as well.

Will review tomorrow afternoon ahead of releasing analytics.js (when I'm not so exhausted), but otherwise look*ing* good to me so far :)

ruemic commented:
Hey so, I've swapped out _user for user(), both seem to work fine, by I guess its better to use the public api :) 

Also, thought about the early return in the identify event, and switched it to only happen if BOTH userId() & anonymousId() are missing, which in theory should never happen, but who knows.

Thanks for the review

mehulsshah commented:
Checking in again on this pull request, is this being picked up in the release/deploy today?

ndhoule commented on lib/blueshift/index.js:
```diff
@@ -89,5 +94,42 @@ Blueshift.prototype.group = function(group){
+ */
+
+function removeBlanks(obj){
+  for (var i in obj) {
```
This needs to either be wrapped in a `hasOwnProperty` check or, better, replace it with `each`

ndhoule commented on lib/blueshift/test.js:
```diff
@@ -94,57 +94,77 @@ describe('Blueshift', function(){
+          anonymousId: analytics.user().anonymousId(),
+          customer_id: analytics._user.id() || undefined,
+        }
+        removeBlanks(properties);
```
Rather than doing dynamic filtering in tests, be explicit about what we're testing here

ndhoule commented on lib/blueshift/test.js:
```diff
@@ -94,57 +94,77 @@ describe('Blueshift', function(){
+        var properties = {
+          _bsft_source: "segment.com",
+          anonymousId: analytics.user().anonymousId(),
+          customer_id: analytics._user.id() || undefined,
```
Why do we need to access a private member here? This isn't part of the analytics API. Is `analytics.user()` insufficient?

ndhoule commented on lib/blueshift/test.js:
```diff
@@ -94,57 +94,77 @@ describe('Blueshift', function(){
   });
 });
+
+function removeBlanks(obj){
```
Making the above test explicit will remove the need for this funciton

ndhoule commented on lib/blueshift/index.js:
```diff
@@ -89,5 +94,42 @@ Blueshift.prototype.group = function(group){
+ */
+
+Blueshift.prototype.alias = function(alias){
+  var properties = {
```
```js
var properties = removeBlanks({ / * ... */ });
```

ndhoule commented on lib/blueshift/index.js:
```diff
@@ -89,5 +94,42 @@ Blueshift.prototype.group = function(group){
+ * @api private
+ */
+
+function removeBlanks(obj){
```
```
var fold = require('foldl');

// ...
function removeBlankAttributes(obj) {
  return foldl(function(results, val, key) {
    if (val !== null && val !== undefined) results[key] = val;
    return results;
  }, {}, obj);
}
```

ndhoule commented on lib/blueshift/index.js:
```diff
@@ -89,5 +94,42 @@ Blueshift.prototype.group = function(group){
   properties._bsft_source = 'segment.com';
+  properties.customer_id = track.userId();
+  properties.anonymousId = track.anonymousId();
+  removeBlanks(properties)
```
missing semicolon

ndhoule commented on lib/blueshift/index.js:
```diff
@@ -89,5 +94,42 @@ Blueshift.prototype.group = function(group){
+    customer_id: alias.userId(),
+    previous_customer_id: alias.previousId(),
+    anonymousId: alias.anonymousId()
+  }
```
missing semicolon

ndhoule commented on lib/blueshift/index.js:
```diff
@@ -89,5 +94,42 @@ Blueshift.prototype.group = function(group){
+    anonymousId: alias.anonymousId()
+  }
+  removeBlanks(properties);
+  window.blueshift.track('alias',properties);
```
space after all unary operators (`,`)

ruemic commented:
@ndhoule thanks for all the excellent feedback, nice little 'foldl' lib!
I've incorporated it here https://github.com/ruemic/analytics.js-integrations/commit/bc809d66cb8f2f527d533ea89d2e8ad02ed3fb58

The the second `track` test still fails because `userId()` returns null in the track stub, and I'm not sure of how to set it in a good way.

For example, I've tried setting it by calling ```analytics.identify('id').track('event');``` instead of ```analytics.track('event');``` before the ```analytics.called(window.blueshift.track, 'event', properties);``` but userId() still is null in the track stub. Any pointers here would be much appreciated.

ruemic commented:
@sperand-io might you have any idea about how we can make the 'should call track event with a userId' work? Otherwise, we can also just remove it...

ruemic commented:
Ok guys, sorry to be such a bother.
We're very eager to get this out so I've simplified the test to just work with the existing stub for 'track'. 
Squashed & rebased as well.

ruemic commented:
I've reworded our test descriptions to be more consistent with other integrations

sperand-io commented:
before merge, just want to confirm that removal of `group()` was intentional. is that the case? @ruemic @mehulsshah

ruemic commented:
@sperand-io yes, we are not using group, thanks for checking

ruemic commented:
I'm rebasing now and fixing errors thrown by what looks like linting changes

ndhoule commented:
@ruemic Thanks for the heads up. I just rebased this myself and corrected linting errors to save you the hassle; I'll be merging shortly. Thanks for the PR!

ndhoule commented:
Moving this over to https://github.com/segmentio/analytics.js-integrations/pull/632, will merge once CI runs.

ruemic commented:
Great, thanks!
